src/client/src/rt-platforms/openFin/adapter/epics.ts

```diff
@@ -1,17 +1,102 @@
 import { Action } from 'redux'
-import { map, switchMapTo } from 'rxjs/operators'
-import { ActionsObservable, ofType, StateObservable } from 'redux-observable'
-import { SetupActions } from 'rt-actions'
-import { LayoutActions } from 'rt-actions/layoutActions'
-import { setupWorkspaces } from './openFin'
+import { fromEventPattern, merge } from 'rxjs'
+import { map, switchMapTo, mergeMap, delay } from 'rxjs/operators'
+import { ActionsObservable, ofType } from 'redux-observable'
+import flowRight from 'lodash/flowRight'
 import { ApplicationEpic } from 'StoreTypes'
-const setupLayout = ofType<Action>(SetupActions.setup)
+import {
+  SetupAction,
+  SETUP_ACTION_TYPES,
+  WORKSPACE_ACTION_TYPES,
+  LAYOUT_ACTION_TYPES,
+  LayoutAction,
+} from 'rt-actions'
+import { setupWorkspaces, restoreWorkspace, saveWorkspace } from './workspaces'
+import { isMainWindow, addApplicationEventHandler } from './window'
+import { WorkspaceActions, WorkspaceAction, LayoutActions } from 'rt-actions'
+
+const DELAY_SAVING_WORKSPACE = 2_000
```

Comment:
Why is there a delay?

Reply:
If you generate the workspaces object immediately after creating/closing/moving/minimising/etc openfin windows,  the object frequently contains empty windows eg url: `about:blank`.

This delay could be reduced, I set it this long to be safe. It's not ideal, but I couldn't reliably get it to work without adding this.

Reply:
Additionally where I use that delay, I've left a comment.